Fabric.Authorization.Persistence.SqlServer/Stores/SqlServerRoleStore.cs

```diff
@@ -118,15 +142,101 @@ public Task<IEnumerable<Role>> GetRoles(string grain, string securableItem = nul
 
             if (!string.IsNullOrWhiteSpace(securableItem))
             {
-                roles = roles.Where(p => string.Equals(p.SecurableItem.Name, securableItem));
+                roles = roles.Where(r => string.Equals(r.SecurableItem.Name, securableItem));
             }
 
             if (!string.IsNullOrWhiteSpace(roleName))
             {
-                roles = roles.Where(p => string.Equals(p.Name, roleName));
+                roles = roles.Where(r => string.Equals(r.Name, roleName));
+            }
+
+            return roles;
+        }
+
+        public async Task<Role> AddPermissionsToRole(Role role, ICollection<Permission> permissions)
+        {
+            var roleEntity = await _authorizationDbContext.Roles
+                .SingleOrDefaultAsync(r =>
+                    r.RoleId == role.Id
+                    && !r.IsDeleted);
+
+            if (role == null)
+            {
+                throw new NotFoundException<Role>($"Could not find {typeof(Role).Name} entity with ID {role.Id}");
+            }
+
+            var permissionIds = permissions.Select(p => p.Id);
+            var permissionDictionary = _authorizationDbContext.Permissions
+                .Where(p => !p.IsDeleted)
+                .Where(p => permissionIds.Contains(p.PermissionId))
+                .ToDictionary(p => p.PermissionId);
+
+            foreach (var permission in permissions)
+            {
+                _authorizationDbContext.RolePermissions.Add(new RolePermission
+                {
+                    RoleId = roleEntity.Id,
+                    PermissionId = permissionDictionary[permission.Id].Id,
+                    PermissionAction = PermissionAction.Allow
+                });
+            }
+
+            await _authorizationDbContext.SaveChangesAsync();
+
+            var updatedRole = await Get(role.Id);
```

Comment:
Do we need to get the role again? We already have a reference to it don't we?

Reply:
I retrieve it again to ensure the Domain model is refreshed correctly before returning. It has collections of Allowed and Denied permissions. I suppose I could pull these entities out of the existing dictionary, transform them to the Permission domain model, and add them to the appropriate collection.

Reply:
Wouldn't the role that was passed just have all the references it needs? I am concerned about the number of queries we are doing in general.

Reply:
The role that is passed in does not yet have the new permissions added. Perhaps I can just add the permission to the passed-in Role's permissions collection. I'll do that to eliminate that query.